Clippy cannot find 'std' with non-standard libdir

When I try even hello world project I got error:

```
$ cargo clippy --verbose
    Checking tst001 v0.1.0 (file:///tmp/tst001)
     Running `/usr/bin/clippy-driver rustc --crate-name tst001 src/main.rs --crate-type bin --emit=dep-info,metadata -C debuginfo=2 -C metadata=9d53d8442a8c6217 -C extra-filename=-9d53d8442a8c6217 --out-dir /tmp/tst001/target/debug/deps -C incremental=/tmp/tst001/target/debug/incremental -L dependency=/tmp/tst001/target/debug/deps`
error[E0463]: can't find crate for `std`

error: Could not compile `tst001`.
```
Clippy version: 0.0.210
Rustc version:
```
rustc 1.28.0-dev (773ce53ce 2018-06-26)
binary: rustc
commit-hash: 773ce53ce7b3acb97cfbd3d189dc3fbf33ec05c6
commit-date: 2018-06-26
host: x86_64-unknown-linux-gnu
release: 1.28.0-dev
LLVM version: 6.0
```

Rustc built with non-standart libdir and libstd placed there:
```
/usr/lib64/rust-9999/libstd-c9650cebf997ec41.so
/usr/lib64/rust-9999/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c9650cebf997ec41.rlib
/usr/lib64/rust-9999/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c9650cebf997ec41.so
```

Rustc successfully builds programs but clippy fails to check it.

--- a/src/driver.rs
+++ b/src/driver.rs
@@ -23,7 +23,7 @@ use self::rustc_driver::{driver::CompileController, Compilation};
 
 use std::convert::TryInto;
 use std::path::Path;
-use std::process::{exit, Command};
+use std::process::exit;
 
 fn show_version() {
     println!(env!("CARGO_PKG_VERSION"));
@@ -40,54 +40,22 @@ pub fn main() {
                 exit(0);
             }
 
-            let sys_root = option_env!("SYSROOT")
-                .map(String::from)
-                .or_else(|| std::env::var("SYSROOT").ok())
-                .or_else(|| {
-                    let home = option_env!("RUSTUP_HOME").or(option_env!("MULTIRUST_HOME"));
-                    let toolchain = option_env!("RUSTUP_TOOLCHAIN").or(option_env!("MULTIRUST_TOOLCHAIN"));
-                    home.and_then(|home| toolchain.map(|toolchain| format!("{}/toolchains/{}", home, toolchain)))
-                })
-                .or_else(|| {
-                    Command::new("rustc")
-                        .arg("--print")
-                        .arg("sysroot")
-                        .output()
-                        .ok()
-                        .and_then(|out| String::from_utf8(out.stdout).ok())
-                        .map(|s| s.trim().to_owned())
-                })
-                .expect("need to specify SYSROOT env var during clippy compilation, or use rustup or multirust");
-
             // Setting RUSTC_WRAPPER causes Cargo to pass 'rustc' as the first argument.
             // We're invoking the compiler programmatically, so we ignore this/
-            let mut orig_args: Vec<String> = env::args().collect();
-            if orig_args.len() <= 1 {
+            let mut args: Vec<String> = env::args().collect();
+            if args.len() <= 1 {
                 std::process::exit(1);
             }
-            if Path::new(&orig_args[1]).file_stem() == Some("rustc".as_ref()) {
+            if Path::new(&args[1]).file_stem() == Some("rustc".as_ref()) {
                 // we still want to be able to invoke it normally though
-                orig_args.remove(1);
+                args.remove(1);
             }
-            // this conditional check for the --sysroot flag is there so users can call
-            // `clippy_driver` directly
-            // without having to pass --sysroot or anything
-            let mut args: Vec<String> = if orig_args.iter().any(|s| s == "--sysroot") {
-                orig_args.clone()
-            } else {
-                orig_args
-                    .clone()
-                    .into_iter()
-                    .chain(Some("--sysroot".to_owned()))
-                    .chain(Some(sys_root))
-                    .collect()
-            };
 
             // this check ensures that dependencies are built but not linted and the final
             // crate is
             // linted but not built
             let clippy_enabled = env::var("CLIPPY_TESTS").ok().map_or(false, |val| val == "true")
-                || orig_args.iter().any(|s| s == "--emit=dep-info,metadata");
+                || args.iter().any(|s| s == "--emit=dep-info,metadata");
 
             if clippy_enabled {
                 args.extend_from_slice(&["--cfg".to_owned(), r#"feature="cargo-clippy""#.to_owned()]);

--- a/tests/compile-test.rs
+++ b/tests/compile-test.rs
@@ -78,7 +78,11 @@ fn config(mode: &str, dir: PathBuf) -> compiletest::Config {
         config.run_lib_path = rustc_lib_path();
         config.compile_lib_path = rustc_lib_path();
     }
-    config.target_rustcflags = Some(format!("-L {0} -L {0}/deps -Dwarnings --sysroot {1}", host_libs().display(), rustc_sysroot_path().display()));
+    config.target_rustcflags = Some(format!(
+        "-L {0} -L {0}/deps -Dwarnings --sysroot {1}",
+        host_libs().display(),
+        rustc_sysroot_path().display()
+    ));
 
     config.mode = cfg_mode;
     config.build_base = if rustc_test_suite().is_some() {

--- a/tests/compile-test.rs
+++ b/tests/compile-test.rs
@@ -17,6 +17,7 @@ use std::ffi::OsStr;
 use std::fs;
 use std::io;
 use std::path::{Path, PathBuf};
+use std::process::Command;
 
 fn clippy_driver_path() -> PathBuf {
     if let Some(path) = option_env!("CLIPPY_DRIVER_PATH") {
@@ -42,6 +43,28 @@ fn rustc_lib_path() -> PathBuf {
     option_env!("RUSTC_LIB_PATH").unwrap().into()
 }
 
+fn rustc_sysroot_path() -> PathBuf {
+    option_env!("SYSROOT")
+        .map(String::from)
+        .or_else(|| std::env::var("SYSROOT").ok())
+        .or_else(|| {
+            let home = option_env!("RUSTUP_HOME").or(option_env!("MULTIRUST_HOME"));
+            let toolchain = option_env!("RUSTUP_TOOLCHAIN").or(option_env!("MULTIRUST_TOOLCHAIN"));
+            home.and_then(|home| toolchain.map(|toolchain| format!("{}/toolchains/{}", home, toolchain)))
+        })
+        .or_else(|| {
+            Command::new("rustc")
+                .arg("--print")
+                .arg("sysroot")
+                .output()
+                .ok()
+                .and_then(|out| String::from_utf8(out.stdout).ok())
+                .map(|s| s.trim().to_owned())
+        })
+        .expect("need to specify SYSROOT env var during clippy compilation, or use rustup or multirust")
+        .into()
+}
+
 fn config(mode: &str, dir: PathBuf) -> compiletest::Config {
     let mut config = compiletest::Config::default();
 
@@ -55,7 +78,7 @@ fn config(mode: &str, dir: PathBuf) -> compiletest::Config {
         config.run_lib_path = rustc_lib_path();
         config.compile_lib_path = rustc_lib_path();
     }
-    config.target_rustcflags = Some(format!("-L {0} -L {0}/deps -Dwarnings", host_libs().display()));
+    config.target_rustcflags = Some(format!("-L {0} -L {0}/deps -Dwarnings --sysroot {1}", host_libs().display(), rustc_sysroot_path().display()));
 
     config.mode = cfg_mode;
     config.build_base = if rustc_test_suite().is_some() {

--- a/tests/dogfood.rs
+++ b/tests/dogfood.rs
@@ -7,6 +7,31 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use std::path::PathBuf;
+use std::process::Command;
+
+fn rustc_sysroot_path() -> PathBuf {
+    option_env!("SYSROOT")
+        .map(String::from)
+        .or_else(|| std::env::var("SYSROOT").ok())
+        .or_else(|| {
+            let home = option_env!("RUSTUP_HOME").or(option_env!("MULTIRUST_HOME"));
+            let toolchain = option_env!("RUSTUP_TOOLCHAIN").or(option_env!("MULTIRUST_TOOLCHAIN"));
+            home.and_then(|home| toolchain.map(|toolchain| format!("{}/toolchains/{}", home, toolchain)))
+        })
+        .or_else(|| {
+            Command::new("rustc")
+                .arg("--print")
+                .arg("sysroot")
+                .output()
+                .ok()
+                .and_then(|out| String::from_utf8(out.stdout).ok())
+                .map(|s| s.trim().to_owned())
+        })
+        .expect("need to specify SYSROOT env var during clippy compilation, or use rustup or multirust")
+        .into()
+}
+
 #[test]
 fn dogfood() {
     if option_env!("RUSTC_TEST_SUITE").is_some() || cfg!(windows) {
@@ -21,6 +46,7 @@ fn dogfood() {
     let output = std::process::Command::new(clippy_cmd)
         .current_dir(root_dir)
         .env("CLIPPY_DOGFOOD", "1")
+        .env("RUSTFLAGS", format!("--sysroot {}", rustc_sysroot_path().display()))
         .arg("clippy")
         .arg("--all-targets")
         .arg("--all-features")
@@ -59,6 +85,7 @@ fn dogfood_tests() {
         let output = std::process::Command::new(&clippy_cmd)
             .current_dir(root_dir.join(d))
             .env("CLIPPY_DOGFOOD", "1")
+            .env("RUSTFLAGS", format!("--sysroot {}", rustc_sysroot_path().display()))
             .arg("clippy")
             .arg("--")
             .args(&["-D", "clippy::all"])

